R/calc_truth.R

```diff
@@ -0,0 +1,328 @@
+#' Calculate "truth" for a stock using {ecosystemdata} outputs
+#' 
+#' @description This function generates standardized "truth" outputs using data 
+#' from {ecosystemdata}. The "truth" include "true" index and age composition
+#' (both monthly and yearly) for a specified stock.
+#' 
+#' @details Currently, this function supports calculating "truth" for biomass, 
+#' catch, numbers, and weight. Numbers-at-age are calculated by dividing 
+#' biomass-at-age by weight-at-age.
+#' 
+#' @param data A tibble containing ecosystem model (e.g., Ecopath with Ecosim) 
+#' outputs from {ecosystemdata}.
+#' @param stock_name A character string specifying the stock name to subset.
+#' 
+#' @return A tibble with nested "truth" (e.g., biomass, catch, 
+#' numbers, or weight), categorized by type (index or age composition) and 
+#' timestep (monthly or yearly).
+#' 
+#' @examples
+#' data(ewe_nwatlantic_base, package = "ecosystemdata")
+#' data <- ewe_nwatlantic_base
+#' truth <- calc_truth(
+#'   data,
+#'   stock_name = "menhaden"
+#' )
+#' @export
+#' @rdname calc_truth
+#' @keywords calc_truth
+calc_truth <- function(data, stock_name){
+
+  # Check if the stock_name is in the data
+  if (!stock_name %in% unique(data[["species"]])) {
+    cli::cli_abort(c(
+      "Stock {.val {stock_name}} not found in the data.",
+      "Please check the stock name and try again."
+    ))
+  }
+
+  # Create a tibble with the labels (e.g., biomass) and types (e.g., index)
+  # Define available labels from data
+  labels <- c("biomass", "catch", "weight")
+  missing_data <- setdiff(labels, unique(data[["type"]]))
+  if (length(missing_data) > 0) {
+    cli::cli_abort(c(
+      "{.val {missing_data}} at age data is missing for {stock_name}.",
+      "Please ensure that the data contains {missing_data} information."
+    ))
+  }
+
+  # Create a tibble that maps labels to their types (index or age composition)
+  types <- dplyr::tibble(
+    label = rep(labels, 2),
+    type = c(
+      # weight only has agecomp type, NA for weight index
+      ifelse(labels %in% c("biomass", "catch"), "index", NA_character_),
+      rep("agecomp", length(labels))
+    )
+  ) |>
+    dplyr::filter(!is.na(type))
+
+  # Extract monthly data by labels
+  truth_monthly <- purrr::map(
+    labels,
+    ~ calc_truth_monthly(data, stock_name, .x)
```

Comment:
The recommendations for purrr suggest
* An anonymous function, e.g. \(x) x + 1 or function(x) x + 1.
* A formula, e.g. ~ .x + 1. You must use .x to refer to the first argument. Only recommended if you require backward compatibility with older versions of R.
Where, the latter is what you are using here but it is only suggested for backward compatibility.